crates/circuit/src/object_registry.rs

```diff
@@ -107,140 +104,150 @@ impl PartialEq for VarAsKey {
+///
+/// If type parameter `B` implements [IntoPyObject], then a cached [PyList]
+/// is maintained and accessible via [ObjectRegistry::cached] and [ObjectRegistry::cached_raw],
+/// which contains the unique objects, in the order they were first registered.
 #[derive(Clone, Debug)]
-pub struct BitData<T, B> {
-    /// Registered Python bits.
-    bits: Vec<B>,
-    /// Maps Python bits to native type.
+pub struct ObjectRegistry<T, B> {
+    /// Registered objects.
+    objects: Vec<B>,
+    /// Maps objects to native index.
     indices: HashMap<B, T>,
-    /// The bits registered, cached as a PyList.
+    /// The objects registered, cached as a PyList.
     cached: OnceLock<Py<PyList>>,
 }
 
-impl<T, B> Default for BitData<T, B>
+impl<T, B> Default for ObjectRegistry<T, B>
 where
-    T: From<BitType> + Copy,
-    BitType: From<T>,
-    B: for<'a> IntoPyObject<'a> + ShareableBit,
+    T: From<u32> + Copy,
+    u32: From<T>,
+    B: Clone + Eq + Hash + Debug,
 {
     fn default() -> Self {
         Self::new()
     }
 }
-impl<T, B> BitData<T, B>
+impl<T, B> ObjectRegistry<T, B>
 where
-    T: From<BitType> + Copy,
-    BitType: From<T>,
-    B: for<'a> IntoPyObject<'a> + ShareableBit,
+    T: From<u32> + Copy,
+    u32: From<T>,
+    B: Clone + Eq + Hash + Debug,
 {
     pub fn new() -> Self {
         Self::with_capacity(0)
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
-        BitData {
-            bits: Vec::with_capacity(capacity),
+        ObjectRegistry {
+            objects: Vec::with_capacity(capacity),
             indices: HashMap::with_capacity(capacity),
             cached: OnceLock::new(),
         }
     }
 
-    /// Gets the number of bits.
+    /// Gets the number of registered objects.
     pub fn len(&self) -> usize {
-        self.bits.len()
+        self.objects.len()
     }
 
     pub fn is_empty(&self) -> bool {
-        self.bits.is_empty()
+        self.objects.is_empty()
     }
 
-    /// Gets a reference to the underlying vector of Python bits.
+    /// Gets a reference to the underlying vector of objects.
     #[inline]
-    pub fn bits(&self) -> &Vec<B> {
-        &self.bits
+    pub fn objects(&self) -> &Vec<B> {
+        &self.objects
     }
 
     /// Gets a reference to the cached Python list, maintained by
     /// this instance.
     #[inline]
-    pub fn cached(&self, py: Python) -> &Py<PyList> {
+    pub fn cached<'py>(&self, py: Python<'py>) -> &Py<PyList>
+    where
+        B: IntoPyObject<'py>,
+    {
         self.cached
-            .get_or_init(|| PyList::new(py, self.bits.clone()).unwrap().into())
+            .get_or_init(|| PyList::new(py, self.objects.clone()).unwrap().into())
     }
 
     /// Gets a reference to the cached Python list, even if not initialized.
     #[inline]
-    pub fn cached_raw(&self) -> Option<&Py<PyList>> {
+    pub fn cached_raw(&self) -> Option<&Py<PyList>>
+    where
+        for<'a> B: IntoPyObject<'a>,
+    {
         self.cached.get()
     }
```

Comment:
Would it make sense to separate these methods into a different impl block with `<B: IntoPyObject<'_>` as an extra trait bound? Since the `PyList` usage is now directly dependent of the object implementing said trait, I belive we could just move this to only work if said condition is met.